pkg/provenance/provenance_test.go

```diff
@@ -0,0 +1,521 @@
+
+	args = deepCopyArgs(args)
+	users2 := addUser(args.Users, "johnson", "fluffy23")
+	args.Users = users2
+	args.Version = 2
+	spec2 := makeSpec(args)
+	objectLineage[spec2.Version] = spec2
+
+	args = deepCopyArgs(args)
+	users3 := addUser(args.Users, "thomas", "bedsheet85")
+	args.Users = users3
+	args.Version = 3
+	spec3 := makeSpec(args)
+	objectLineage[spec3.Version] = spec3
+
+	args = deepCopyArgs(args)
+	db1 := addDatabase(args.Databases, "logging")
+	args.Databases = db1
+	args.Version = 4
+	spec4 := makeSpec(args)
+	objectLineage[spec4.Version] = spec4
+
+	args = deepCopyArgs(args)
+	db2 := addDatabase(args.Databases, "demographics")
+	args.Databases = db2
+	args.Version = 5
+	spec5 := makeSpec(args)
+	objectLineage[spec5.Version] = spec5
+	args = deepCopyArgs(args)
+	return objectLineage, args
+}
+
+// This method is testing that the joint functionality of Bisect works
+// with a complex query involving a complex query of 3 fields
+func TestBisectGeneral(t *testing.T) {
+	objLineage, _ := buildLineage()
+	argMapTest := make(map[string]string, 0)
+	argMapTest["field1"] = "databases"
+	argMapTest["value1"] = "demographics"
+	argMapTest["field2"] = "username"
+	argMapTest["value2"] = "daniel"
+	argMapTest["field3"] = "password"
+	argMapTest["value3"] = "4557832+#^"
+
+	vOutput := objLineage.Bisect(argMapTest)
+	expected := "Version: 5"
+	if vOutput != expected {
+		t.Errorf("Version output for TestBisectGeneral() was incorrect, got: %s, want: %s.\n", vOutput, expected)
+	}
+}
+// This method changes a user's password, and is testing
+// whether the bisect will take a query with the same
+// parent element (user), and process a complicated
+// query involving maps.
+func TestBisectPasswordChanged(t *testing.T) {
+	objLineage, newArgs := buildLineage()
+
+	users := updateUser(newArgs.Users, "daniel", "JEK873BUL!")
+	db := removeDatabase(newArgs.Databases, "logging")
+	newArgs.Users = users
+	newArgs.Databases = db
+	newArgs.Version = 6
+	spec6 := makeSpec(newArgs)
+	objLineage[spec6.Version] = spec6
+	argMapTest := make(map[string]string, 0)
+	argMapTest["field1"] = "username"
+	argMapTest["value1"] = "daniel"
+	argMapTest["field2"] = "password"
+	argMapTest["value2"] = "JEK873BUL!"
+
+	vOutput := objLineage.Bisect(argMapTest)
+	expected := "Version: 6"
+	if vOutput != expected {
+		t.Errorf("Version output for TestBisectPasswordChanged() was incorrect, got: %s, want: %s.\n", vOutput, expected)
+	}
+}
+// This method removes a databases, and
+// is testing that Bisect still works.
+func TestBisectRemoveDatabase(t *testing.T) {
+	objLineage, newArgs := buildLineage()
+	args := deepCopyArgs(newArgs)
+
+	db := removeDatabase(args.Databases, "logging")
+	args.Databases = db
+	args.Version = 6
+	spec6 := makeSpec(args)
+	objLineage[spec6.Version] = spec6
+
+	argMapTest := make(map[string]string, 0)
+	argMapTest["field1"] = "databases"
+	argMapTest["value1"] = "logging"
+
+	vOutput := objLineage.Bisect(argMapTest)
+	expected := "Version: 4"
```

Comment:
Shouldn't the expected be "Version: 6" ?

Reply:
I think I understand. Essentially we are saying that if an attribute is removed then Bisect will show the latest version in which that attribute existed. Correct?